python/hsfs/engine/spark.py

```diff
@@ -315,14 +315,41 @@ def write_training_dataset(
 
         dataset = self.convert_to_default_dataframe(dataset)
 
-        # apply transformation functions
+        # generate transformation function expressions
+        transformed_feature_names = []
+        transformation_fn_expressions = []
         for (
             feature_name,
             transformation_fn,
         ) in training_dataset.transformation_functions.items():
-            dataset = dataset.withColumn(
-                feature_name, transformation_fn.transformation_fn(col(feature_name))
+            fn_registration_name = (
+                transformation_fn.name + "_" + str(transformation_fn.version)
             )
+            self._spark_session.udf.register(
+                fn_registration_name, transformation_fn.transformation_fn
+            )
+            transformation_fn_expressions.append(
+                "{fn_name:}({name:}) AS {name:}".format(
```

Comment:
why do you need the `:` I think they add some formatting spaces in some cases, is that wanted?

Reply:
yeah, it doesn't make any difference. It seemed better readable for my eyes :)